evm/vm/base.py

```diff
@@ -122,6 +133,49 @@ def get_block_reward(self, block_number):
     def get_nephew_reward(self, block_number):
         return NEPHEW_REWARD
 
+    def import_block(self, block):
+        parent_header = self.evm.get_block_header_by_hash(block.header.parent_hash)
+        init_header = self.create_header_from_parent(parent_header)
+        evm = type(self.evm)(self.db, init_header)
+        vm = type(self)(evm, self.db)
```

Comment:
Just a comment as I'm not sure how to fix it but:

This is indicative that I've gotten an abstraction wrong.  The shared header object across the `evm`, `vm`, and `block` is making for awkward code like this.  It's not clear to me how to fix it but I think it's something that *needs* to eventually be sorted out.  If you come up with any ideas I'm all ears.

Here are some extra thoughts that may or may not help solve this.

* The `EVM` class uses the header to know the *current* block number and thus, which `VM` class needs to be used.
* The `VM` class uses the header to initialize its `self.block` instance.
* When mining a new block, the `EVM` class uses the header to initialized the header for the next block.
* The `EVM.header` property gets mutated in the methods `EVM.apply_transaction()`, `EVM.import_block()`, `EVM.mine_block()`, and `EVM.configure_header()`.  Mutation is inherently problematic in this situation because both the `EVM` and `VM` classes use the header as an anchor for the current chain state and since it gets mutated, it opens up ways for us to have two components with different opinions on what the current chain state should be.
* `EVM.from_genesis()` has awkward code at the end which first initializes the `EVM` instance with the genesis header and then replaces it with a new header initialized from the genesis header.

I suspect we're not looking at one architecture problem but 2 or 3 that are combining into something more hairy.